L29-opencensus-filter.md

```diff
@@ -0,0 +1,40 @@
+Title
+----
+* Author(s): Jim King, Emil Mukilic
+* Approver:
+* Status: In Review
+* Implemented in: C++
+* Last updated: 5-14-2018
+* Discussion at:
+
+## Abstract
+
+The OpenCensus filter allows grpc to itegrate and collect stats and tracing data with OpenCensus (https://github.com/census-instrumentation/opencensus-cpp). The Opencensus filter uses a number of internal APIs (src/core/lib/slice/slice_internal.h, src/cpp/common/channel_filter.h, src/core/lib/gprpp/orphanable.h, src/core/lib/surface/call.h). Since these internal APIs are subject to change, in order to ease maintenance of the filter, we propose that the filter be moved to the grpc repository (src/cpp/ext/filters/census/). This will introduce an API change for grpc++ when OpenCensus is used, which will require the user to register the OpenCensus filter with grpc.  This is in addition to any views and exporters that would need to be registered to use OpenCensus.
+
+## Background
+
+OpenCensus tracing and stats functionality is a desired feature for grpc C++ users. Integrating OpenCensus into grpc core as a direct dependency is not viable due to OpenCensus dependencies which are not supported in grpc (abseil being a primary one).
+
+### Related Proposals:
+N/A
+
+## Proposal
+
+We propose that the OpenCensus filter (C++ grpc filter) which currently resides in the OpenCensus repository (https://github.com/census-instrumentation/opencensus-cpp) be moved to the grpc repository (src/cpp/ext/filters/census/). The OpenCensus filter will be setup as an optional build target which users can include.  There no viable way to include it by default with the default grpc build due to dependency conflicts. Users will have to manually enable the filter by using a filter registration call.
+
+The major changes introduced by this are as follows:
+  1) A new optional dependency on OpenCensus. This will be a new build target in the Bazel build file. Currently only the Bazel build will be supported.  There will be no git submodule for OpenCensus, and no make/cmake support (this may be added at a later time).
+  2) A new API call will be introduced to enable the OpenCensus filter for tracing and stats collection. This must be called to register the plugin mechanism and initialize OpenCensus. RegisterOpenCensusPlugin() will reside in src/cpp/ext/filters/census/grpc_plugin.h.  There will also be an optional API call to register default views for OpenCensus.
+
```

Comment:
Since this lives in `src,` is the intention that this will be available only to users who build their targets with bazel? If so, this is a stronger statement than what's described in the bullet above (which makes it seem like it's just the library that requires building with bazel). I know it's not currently a thing, but the bazel team has an end of 2018 roadmap for building a fully static library with bazel (linked from bazelbuild/bazel#1920) and I suppose that that could be linked to a target using other tools.

Reply:
Yes, at the moment it will only be available to users building with Bazel.

Reply:
After that change from the bazel team happens, it will probably be possible to build it as a static library and link it with another set of build targets that were not built with bazel.